File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

 /*

 */
 
public class Solution {
    public ListNode ReverseList(ListNode head) {
        // Initialize prev variable to null and curr to point at head
        ListNode prev = null, curr = head;

        // While the current node is not null
        while (curr != null) {
            // Create a temporary variable to hold the node that is next from curr
            var temp = curr.next;

            // Assign the next curr value to be whatever was previous
            curr.next = prev;

            // Move prev to the current value
            prev = curr;

            // Assign current to the temporary value
            curr = temp;
        }

        // Once outside of the while loop prev will be the reversed list
        return prev;
    }
}

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        old_to_cur = {None : None}

        cur = head
        while cur:
            copy = Node(cur.val)
            old_to_cur[cur] = copy
            cur = cur.next

        cur = head
        while cur:
            copy = old_to_cur[cur]
            copy.next = old_to_cur[cur.next]
            copy.random = old_to_cur[cur.random]
            cur = cur.next

        return old_to_cur[head]